src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CompletionSource.cs

```diff
@@ -234,12 +226,134 @@ static void CheckForExperimentStatus(ITextView textView, in CompletionOptions op
+                AddPropertiesToSession(session, list, triggerLocation);
+                return context;
+            }
+
+            // The use of an explicit task (instead of calling GetCompletionContextWorkerAsync w/o await) is to
+            // make sure the entire expanded item computation runs in the background.
+            var backgroundTaskCancellationToken = GetCancellationTokenForSession(session);
+            var expandedItemsTask = Task.Run(async () => await GetCompletionContextWorkerAsync(document, trigger, triggerLocation,
+                options with { ExpandedCompletionBehavior = ExpandedCompletionMode.ExpandedItemsOnly }, backgroundTaskCancellationToken).ConfigureAwait(false),
+                backgroundTaskCancellationToken);
+
+            // Record how long it takes for the background task to complete.
+            var stopwatch = Stopwatch.StartNew();
+            expandedItemsTask = expandedItemsTask.ContinueWith(task =>
+            {
+                AsyncCompletionLogger.LogAdditionalTicksToCompleteDelayedImportCompletionDataPoint((int)stopwatch.ElapsedMilliseconds);
+                return task.Result;
+            }, backgroundTaskCancellationToken, TaskContinuationOptions.OnlyOnRanToCompletion, TaskScheduler.Default);
+
+            // Now trigger and wait for core providers to return;
+            var (nonExpandedContext, nonExpandedCompletionList) = await GetCompletionContextWorkerAsync(document, trigger, triggerLocation,
+                    options with { ExpandedCompletionBehavior = ExpandedCompletionMode.NonExpandedItemsOnly }, cancellationToken).ConfigureAwait(false);
+            AddPropertiesToSession(session, nonExpandedCompletionList, triggerLocation);
+
+            if (!expandedItemsTask.IsCompleted && !options.BlockOnExpandedCompletion)
+            {
+                // expanded item task still running. Save it to the session and return non-expanded items immediately.
+                session.Properties[ExpandedItemsTask] = expandedItemsTask;
+                AsyncCompletionLogger.LogImportCompletionGetContext(isBlocking: false, delayed: true);
+
+                return nonExpandedContext;
+            }
+
+            AsyncCompletionLogger.LogImportCompletionGetContext(isBlocking: options.BlockOnExpandedCompletion, delayed: false);
+
+            // the task of expanded item is completed, so get the result and combine it with result of non-expanded items.
+            var (expandedContext, expandedCompletionList) = await expandedItemsTask.ConfigureAwait(false);
+            AddPropertiesToSession(session, expandedCompletionList, triggerLocation: null);
+
+            return CombineCompletionContext(nonExpandedContext, expandedContext);
+
+            // Create a cancellation token associated with the session for the background expanded item task.
+            // TODO: not sure if we need a separate token, need to confirm.
```

Comment:
I'm talking to editor folks about this... will update the PR once I have an answer

Reply:
based on what I heard from @olegtk, seems we don't need a separate CancellationToken for this background task tie to a specific completion session, the same token passed in by editor would only be cancelled when a session is done (dismiss or complete). Is that correct? Also tag @AmadeusW just in case :)